### PR TITLE
Feature/ssl

### LINKED ISSUE #2587: [QE]Crypto in transportation

### PR DESCRIPTION
**Pull request by Shylock-Hg**
See detail in #2556 , the client should be implemented by corresponding maintainer @vesoft-inc/graph-reviewers 

The test certificates generating require interactive, and almost other open source projects like even openssl also put the test certificates to source code, so we follow them.

Close #2587

### CODE REVIEW

File: src/common/ssl/SSLConfig.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+#include "common/ssl/SSLConfig.h"
+
+DEFINE_string(cert_path, "", "Path to cert pem.");
+DEFINE_string(key_path, "", "Path to cert key.");
+DEFINE_string(ca_path, "", "Path to trusted CA file.");
+DEFINE_bool(enable_ssl, false, "Wether enable ssl.");
+DEFINE_bool(enable_graph_ssl, false, "Wether enable ssl of graph server.");
+DEFINE_bool(enable_meta_ssl, false, "Wether enable ssl of meta server.");
```
Comment (Aiee) [inline src/common/ssl/SSLConfig.cpp:14]: Change "Wether" to "Whether"

File: src/common/thrift/ThriftClientManager.h [hunk 2]
```diff
@@ -34,6 +37,8 @@ class ThriftClientManager final {
                                        >;
 
   folly::ThreadLocal<ClientMap> clientMap_;
+  // wether enable ssl
```
Comment (Aiee) [inline src/common/thrift/ThriftClientManager.h:40]: ditto

File: tests/nebula-test-run.py [hunk 3]
```diff
@@ -52,6 +52,22 @@ def init_parser():
...
+                          default=False,
+                          help='Wether enable SSL for graph server.')
+    opt_parser.add_option('--enable_meta_ssl',
+                          dest='enable_meta_ssl',
+                          default=False,
+                          help='Wether enable SSL for meta server.')
+    opt_parser.add_option('--ca_signed',
+                          dest='ca_signed',
+                          default=False,
+                          help='Wether enable CA signed SSL/TLS mode.')
```
Comment (Aiee) [inline tests/nebula-test-run.py:70]: ditto

File: src/common/ssl/SSLConfig.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/ssl/SSLConfig.cpp]: Good catch.

File: src/daemons/GraphDaemon.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <errno.h>
+#include <folly/ssl/Init.h>
```
[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/daemons/GraphDaemon.cpp:8]: ok

File: src/common/thrift/ThriftClientManager-inl.h [hunk 5]
```diff
@@ -71,9 +73,16 @@ std::shared_ptr<ClientType> ThriftClientManager<ClientType>::client(const HostAd
...
+  if (enableSSL_) {
+    socket = folly::AsyncSSLSocket::newSocket(nebula::createSSLContext(), evb);
+  }
+  evb->runImmediatelyOrRunInEventBaseThreadAndWait([this, &socket, evb, resolved]() {
+    if (enableSSL_) {
+      socket->connect(nullptr, resolved.host, resolved.port, FLAGS_conn_timeout_ms);
+    } else {
+      socket =
+          folly::AsyncSocket::newSocket(evb, resolved.host, resolved.port, FLAGS_conn_timeout_ms);
+    }
```
Comment (yixinglu) [inline src/common/thrift/ThriftClientManager-inl.h:85]: why do we need to create the socket inside `runImmediatelyOrRunInEventBaseThreadAndWait`? why not do this like following logics:

```c++
    if (enableSSL_) {
      socket = folly::AsyncSSLSocket::newSocket(nebula::createSSLContext(), evb);
      socket->connect(nullptr, resolved.host, resolved.port, FLAGS_conn_timeout_ms);
    } else {
      socket =  folly::AsyncSocket::newSocket(evb, resolved.host, resolved.port, FLAGS_conn_timeout_ms);
    }
```

**Review by yixinglu — COMMENTED** [review]
Please do not directly commit some private keys and certificates to the repo, you can use scripts or Makefile to generate them as needed

File: src/common/thrift/ThriftClientManager-inl.h (hunk 5 above)
Comment (Shylock-Hg) [reply src/common/thrift/ThriftClientManager-inl.h:83]: Good catch. Fixed.

**Comment by Shylock-Hg** [comment]
> Please do not directly commit some private keys and certificates to the repo, you can use scripts or Makefile to generate them as needed

The test certificates generating require interaction, and almost other open source projects like even openssl also put the test certificates to source code, so we follow them.

**Review by Aiee — DISMISSED** [review]

File: src/common/ssl/SSLConfig.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "common/ssl/SSLConfig.h"
+
+DEFINE_string(cert_path, "", "Path to cert pem.");
+DEFINE_string(key_path, "", "Path to cert key.");
+DEFINE_string(ca_path, "", "Path to trusted CA file.");
+DEFINE_bool(enable_ssl, false, "Whether enable ssl.");
```
Comment (yixinglu) [inline src/common/ssl/SSLConfig.cpp:12]: whether to enable ssl

File: src/common/ssl/SSLConfig.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+DEFINE_bool(enable_graph_ssl, false, "Whether enable ssl of graph server.");
+DEFINE_bool(enable_meta_ssl, false, "Whether enable ssl of meta server.");
```
Comment (yixinglu) [inline src/common/ssl/SSLConfig.cpp:14]: it seems that two flags is enough to cover the cases about inside and outside the nebula cluster. Do we really need to add another meta_ssl for users to make a decision?

File: .github/workflows/pull_request.yml [hunk 8]
```diff
@@ -127,7 +127,24 @@ jobs:
...
+              ;;
+            ubuntu2004)
+              # ssl cluster
+              make ENABLE_SSL=true up
+              ;;
+            esac
+            ;;
+          clang-*)
+            # graph ssl only cluster
+            make ENABLE_SSL=false ENABLE_GRAPH_SSL=true up
```
Comment (yixinglu) [inline .github/workflows/pull_request.yml:145]: It seems that there is no test for CA_SIGNED.

File: src/common/ssl/SSLConfig.cpp (hunk 7 above)
Comment (Shylock-Hg) [reply src/common/ssl/SSLConfig.cpp:14]: It's the suggestion from @sherman-the-tank , which to encrypt the sensitive data in meta.

File: .github/workflows/pull_request.yml (hunk 8 above)
Comment (Shylock-Hg) [reply .github/workflows/pull_request.yml:145]: Good catch. Fixed.

File: src/common/ssl/SSLConfig.cpp (hunk 6 above)
Comment (Shylock-Hg) [reply src/common/ssl/SSLConfig.cpp]: ok

**Review by jievince — APPROVED** [review]
LGTM

**Review by Aiee — APPROVED** [review]

File: src/common/ssl/SSLConfig.cpp (hunk 7 above)
Comment (yixinglu) [reply src/common/ssl/SSLConfig.cpp]: I mean whether `enable_ssl` and `enable_graphd_ssl` can already cover the `enable_meta_ssl` scenario, just like below:

```
* enable_graphd_ssl=true & enable_ssl=false

clients <- ssl -> graphd <- no ssl -> metad <- no ssl -> storaged
                     ^                                    ^
                     |______________ no ssl ______________|


* enable_graphd_ssl=false & enable_ssl=true

clients <- no ssl -> graphd <- ssl -> metad <- ssl -> storaged
                       ^                                 ^
                       |______________ ssl ______________|


* enable_graphd_ssl=true & enable_ssl=true

clients <- ssl -> graphd <- ssl -> metad <- ssl -> storaged
                    ^                                 ^
                    |______________ ssl ______________|

* enable_graphd_ssl=false & enable_ssl=false

clients <- no ssl -> graphd <- no ssl -> metad <- no ssl -> storaged
                        ^                                    ^
                        |______________ no ssl ______________|
```

Is `enable_meta_ssl` just to encrypt schema information when communicating with meta? If so, make sense.

**Pull request merged**